### PR TITLE
Call zero grad on each train step

### PR DESCRIPTION
**Pull request by BenjaminBossan**
This fixes a bug introduced by moving the `optimizer.zero_grad()` call
outside of the train step function, making it incompatible with LBFGS
and other optimizers that call the train step several times per
batch and expect the gradient to be reset after each call.

I extended the existing test for LBFGS to actually check the
training and validation scores after the fit call. With the
faulty behavior, the scores are actually diverging, and with the
bugfix, we get reasonable scores.

In #636, I wrote

> So we could introduce a parameter like perform_zero_grad_each_step=True or so (not a very nice name) and if a user wants to do their custom stuff with gradients, they need to turn it off.

After some thinking, I believe it's not actually necessary to
introduce an argument to turn off the `zero_grad` call during
each train step. The reason is that if a user wants to do some
custom stuff like gradient accumulation, they'll have to override
`train_step` anyway, which means they'll control how exactly
`zero_grad` will be called.

### CODE REVIEW

File: skorch/tests/test_net.py [hunk 1]
```diff
@@ -2138,14 +2138,22 @@ def test_batch_count(self, net_cls, module_cls, data, batch_size):
         assert net.history[:, "train_batch_count"] == [train_batch_count]
         assert net.history[:, "valid_batch_count"] == [valid_batch_count]
 
+    @flaky(max_runs=3)
```
Comment (thomasjpfan) [inline skorch/tests/test_net.py:2141]: Does increasing `max_epochs` make this less flaky?

Comment (BenjaminBossan) [reply skorch/tests/test_net.py:2141]: The test is not actually flaky. I just added it because I thought it to be cautious for any test where we expect the model to reach a certain score.

My general impression is that tests have become less flaky, perhaps because of the `seeds_fixed` auto-fixture. Do you believe we should remove the flaky decorators?

Comment (thomasjpfan) [reply skorch/tests/test_net.py:2141]: It would be nice to have a `seeds_fixed`. Although this is not a blocker for this PR.

Comment (BenjaminBossan) [reply skorch/tests/test_net.py:2141]: We have it already here:

https://github.com/skorch-dev/skorch/blob/9fafede929891e1c5c98ba4e7764035b75a25d26/skorch/tests/conftest.py#L23-L27

Comment (thomasjpfan) [reply skorch/tests/test_net.py:2141]: I was just about to implement this and saw it. You added it not too long ago. :) This should mean our tests are deterministic. (hopefully)

**Comment by BenjaminBossan** [comment]
ping @ottonemo
